docs/man/man8/reframe.settings.8
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "REFRAME.SETTINGS" "8" "Sep 29, 2022" "4.0.0-dev.1" "ReFrame"
.SH NAME
reframe.settings \- ReFrame Configuration Manual
.sp
ReFrame’s behavior can be configured through its configuration file (see \fI\%Configuring ReFrame for Your Site\fP), environment variables and command\-line options.
An option can be specified via multiple paths (e.g., a configuration file parameter and an environment variable), in which case command\-line options precede environment variables, which in turn precede configuration file options.
This section provides a complete reference guide of the configuration options of ReFrame that can be set in its configuration file or specified using environment variables.
.sp
ReFrame’s configuration is in JSON syntax.
The full schema describing it can be found in \fI\%reframe/schemas/config.json\fP file.
Any configuration file given to ReFrame is validated against this schema.
.sp
The syntax we use in the following to describe the different configuration object attributes is a valid query string for the \fI\%jq(1)\fP command\-line processor.
.SH TOP-LEVEL CONFIGURATION
.sp
The top\-level configuration object is essentially the full configuration of ReFrame.
It consists of the following properties:
.INDENT 0.0
.TP
.B \&.systems
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
A list of \fI\%system configuration objects\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.environments
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
A list of \fI\%environment configuration objects\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.logging
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
A list of \fI\%logging configuration objects\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.schedulers
.INDENT 7.0
.TP
.B Required
No
.UNINDENT
.sp
A list of \fI\%scheduler configuration objects\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.modes
.INDENT 7.0
.TP
.B Required
No
.UNINDENT
.sp
A list of \fI\%execution mode configuration objects\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.general
.INDENT 7.0
.TP
.B Required
No
.UNINDENT
.sp
A list of \fI\%general configuration objects\fP\&.
.UNINDENT
.SH SYSTEM CONFIGURATION
.INDENT 0.0
.TP
.B \&.systems[]\&.name
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
The name of this system.
Only alphanumeric characters, dashes (\fB\-\fP) and underscores (\fB_\fP) are allowed.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.descr
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB""\fP
.UNINDENT
.sp
The description of this system.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.hostnames
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
A list of hostname regular expression patterns in Python \fI\%syntax\fP, which will be used by the framework in order to automatically select a system configuration.
For the auto\-selection process, see \fI\%here\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.max_local_jobs
The maximum number of forced local build or run jobs allowed.
.sp
Forced local jobs run within the execution context of ReFrame.
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB8\fP
.UNINDENT
.sp
New in version 3.10.0.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.modules_system
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB"nomod"\fP
.UNINDENT
.sp
The modules system that should be used for loading environment modules on this system.
Available values are the following:
.INDENT 7.0
.IP \(bu 2
\fBtmod\fP: The classic Tcl implementation of the \fI\%environment modules\fP (version 3.2).
.IP \(bu 2
\fBtmod31\fP: The classic Tcl implementation of the \fI\%environment modules\fP (version 3.1).
A separate backend is required for Tmod 3.1, because Python bindings are different from Tmod 3.2.
.IP \(bu 2
\fBtmod32\fP: A synonym of \fBtmod\fP\&.
.IP \(bu 2
\fBtmod4\fP: The \fI\%new environment modules\fP implementation (versions older than 4.1 are not supported).
.IP \(bu 2
\fBlmod\fP: The \fI\%Lua implementation\fP of the environment modules.
.IP \(bu 2
\fBspack\fP: \fI\%Spack\fP’s built\-in mechanism for managing modules.
.IP \(bu 2
\fBnomod\fP: This is to denote that no modules system is used by this system.
.UNINDENT
.sp
New in version 3.4: The \fBspack\fP backend is added.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.modules
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list of \fI\%environment module objects\fP to be loaded always when running on this system.
These modules modify the ReFrame environment.
This is useful in cases where a particular module is needed, for example, to submit jobs on a specific system.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.variables
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list of environment variables to be set always when running on this system.
These variables modify the ReFrame environment.
Each environment variable is specified as a two\-element list containing the variable name and its value.
You may reference other environment variables when defining an environment variable here.
ReFrame will expand its value.
Variables are set after the environment modules are loaded.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.prefix
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB"."\fP
.UNINDENT
.sp
Directory prefix for a ReFrame run on this system.
Any directories or files produced by ReFrame will use this prefix, if not specified otherwise.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.stagedir
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB"${RFM_PREFIX}/stage"\fP
.UNINDENT
.sp
Stage directory prefix for this system.
This is the directory prefix, where ReFrame will create the stage directories for each individual test case.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.outputdir
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB"${RFM_PREFIX}/output"\fP
.UNINDENT
.sp
Output directory prefix for this system.
This is the directory prefix, where ReFrame will save information about the successful tests.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.resourcesdir
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB"."\fP
.UNINDENT
.sp
Directory prefix where external test resources (e.g., large input files) are stored.
You may reference this prefix from within a regression test by accessing the \fI\%reframe.core.systems.System.resourcesdir\fP attribute of the current system.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
A list of \fI\%system partition configuration objects\fP\&.
This list must have at least one element.
.UNINDENT
.SS System Partition Configuration
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.name
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
The name of this partition.
Only alphanumeric characters, dashes (\fB\-\fP) and underscores (\fB_\fP) are allowed.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.descr
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB""\fP
.UNINDENT
.sp
The description of this partition.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.scheduler
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
The job scheduler that will be used to launch jobs on this partition.
Supported schedulers are the following:
.INDENT 7.0
.IP \(bu 2
\fBlocal\fP: Jobs will be launched locally without using any job scheduler.
.IP \(bu 2
\fBoar\fP: Jobs will be launched using the \fI\%OAR\fP scheduler.
.IP \(bu 2
\fBpbs\fP: Jobs will be launched using the \fI\%PBS Pro\fP scheduler.
.IP \(bu 2
\fBsge\fP: Jobs will be launched using the \fI\%Sun Grid Engine\fP scheduler.
.IP \(bu 2
\fBslurm\fP: Jobs will be launched using the \fI\%Slurm\fP scheduler.
This backend requires job accounting to be enabled in the target system.
If not, you should consider using the \fBsqueue\fP backend below.
.IP \(bu 2
\fBsqueue\fP: Jobs will be launched using the \fI\%Slurm\fP scheduler.
This backend does not rely on job accounting to retrieve job statuses, but ReFrame does its best to query the job state as reliably as possible.
.IP \(bu 2
\fBtorque\fP: Jobs will be launched using the \fI\%Torque\fP scheduler.
.IP \(bu 2
\fBlsf\fP: Jobs will be launched using the \fI\%LSF\fP scheduler.
.UNINDENT
.sp
New in version 3.7.2: Support for the SGE scheduler is added.

.sp
New in version 3.8.2: Support for the OAR scheduler is added.

.sp
New in version 3.11.0: Support for the LSF scheduler is added.

.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
The way that multiple node jobs are submitted using the SGE scheduler can be very site\-specific.
For this reason, the \fBsge\fP scheduler backend does not try to interpret any related arguments, e.g., \fBnum_tasks\fP, \fBnum_tasks_per_node\fP etc.
Users must specify how these resources are to be requested by setting the \fBresources\fP partition configuration parameter and then request them from inside a test using the \fI\%extra_resources\fP test attribute.
Here is an example configuration for a system partition named \fBfoo\fP that defines different ways for submitting MPI\-only, OpenMP\-only and MPI+OpenMP jobs:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
{
    \(aqname\(aq: \(aqfoo\(aq,
    \(aqscheduler\(aq: \(aqsge\(aq,
    \(aqresources\(aq: [
        {
            \(aqname\(aq: \(aqsmp\(aq,
            \(aqoptions\(aq: [\(aq\-pe smp {num_slots}\(aq]
        },
        {
            \(aqname\(aq: \(aqmpi\(aq,
            \(aqoptions\(aq: [\(aq\-pe mpi {num_slots}\(aq]
        },
        {
            \(aqname\(aq: \(aqmpismp\(aq,
            \(aqoptions\(aq: [\(aq\-pe mpismp {num_slots}\(aq]
        }
    ]
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Each test then can request the different type of slots as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
self.extra_resouces = {
    \(aqsmp\(aq: {\(aqnum_slots\(aq: self.num_cpus_per_task},
    \(aqmpi\(aq: {\(aqnum_slots\(aq: self.num_tasks},
    \(aqmpismp\(aq: {\(aqnum_slots\(aq: self.num_tasks*self.num_cpus_per_task}
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Notice that defining \fI\%extra_resources\fP does not make the test non\-portable to other systems that have different schedulers;
the \fBextra_resources\fP will be simply ignored in this case and the scheduler backend will interpret the different test fields in the appropriate way.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.launcher
.INDENT 7.0
.TP
.B Required
Yes
.UNINDENT
.sp
The parallel job launcher that will be used in this partition to launch parallel programs.
Available values are the following:
.INDENT 7.0
.IP \(bu 2
\fBalps\fP: Parallel programs will be launched using the \fI\%Cray ALPS\fP \fBaprun\fP command.
.IP \(bu 2
\fBibrun\fP: Parallel programs will be launched using the \fBibrun\fP command.
This is a custom parallel program launcher used at \fI\%TACC\fP\&.
.IP \(bu 2
\fBlocal\fP: No parallel program launcher will be used.
The program will be launched locally.
.IP \(bu 2
\fBlrun\fP: Parallel programs will be launched using \fI\%LC Launcher\fP’s \fBlrun\fP command.
.IP \(bu 2
\fBlrun\-gpu\fP: Parallel programs will be launched using \fI\%LC Launcher\fP’s \fBlrun \-M "\-gpu"\fP command that enables the CUDA\-aware Spectrum MPI.
.IP \(bu 2
\fBmpirun\fP: Parallel programs will be launched using the \fBmpirun\fP command.
.IP \(bu 2
\fBmpiexec\fP: Parallel programs will be launched using the \fBmpiexec\fP command.
.IP \(bu 2
\fBsrun\fP: Parallel programs will be launched using \fI\%Slurm\fP’s \fBsrun\fP command.
.IP \(bu 2
\fBsrunalloc\fP: Parallel programs will be launched using \fI\%Slurm\fP’s \fBsrun\fP command, but job allocation options will also be emitted.
This can be useful when combined with the \fBlocal\fP job scheduler.
.IP \(bu 2
\fBssh\fP: Parallel programs will be launched using SSH.
This launcher uses the partition’s \fI\%access\fP property in order to determine the remote host and any additional options to be passed to the SSH client.
The ssh command will be launched in “batch mode,” meaning that password\-less access to the remote host must be configured.
Here is an example configuration for the ssh launcher:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
{
    \(aqname\(aq: \(aqfoo\(aq
    \(aqscheduler\(aq: \(aqlocal\(aq,
    \(aqlauncher\(aq: \(aqssh\(aq
    \(aqaccess\(aq: [\(aq\-l admin\(aq, \(aqremote.host\(aq],
    \(aqenvirons\(aq: [\(aqbuiltin\(aq],
}
.ft P
.fi
.UNINDENT
.UNINDENT
.IP \(bu 2
\fBupcrun\fP: Parallel programs will be launched using the \fI\%UPC\fP \fBupcrun\fP command.
.IP \(bu 2
\fBupcxx\-run\fP: Parallel programs will be launched using the \fI\%UPC++\fP \fBupcxx\-run\fP command.
.UNINDENT
.sp
\fBTIP:\fP
.INDENT 7.0
.INDENT 3.5
New in version 4.0.0: ReFrame also allows you to register your own custom launchers simply by defining them in the configuration.
You can follow a small tutorial \fI\%here\fP\&.

.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.access
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list of job scheduler options that will be passed to the generated job script for gaining access to that logical partition.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.environs
.INDENT 7.0
.INDENT 3.5
.INDENT 0.0
.TP
.B required
No
.TP
.B default
\fB[]\fP
.UNINDENT
.UNINDENT
.UNINDENT
.sp
A list of environment names that ReFrame will use to run regression tests on this partition.
Each environment must be defined in the \fI\%environments\fP section of the configuration and the definition of the environment must be valid for this partition.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.container_platforms
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list for \fI\%container platform configuration objects\fP\&.
This will allow launching regression tests that use containers on this partition.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.modules
.INDENT 7.0
.INDENT 3.5
.INDENT 0.0
.TP
.B required
No
.TP
.B default
\fB[]\fP
.UNINDENT
.UNINDENT
.UNINDENT
.sp
A list of \fI\%environment module objects\fP to be loaded before running a regression test on this partition.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.time_limit
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fBnull\fP
.UNINDENT
.sp
The time limit for the jobs submitted on this partition.
When the value is \fBnull\fP, no time limit is applied.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.variables
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list of environment variables to be set before running a regression test on this partition.
Each environment variable is specified as a two\-element list containing the variable name and its value.
You may reference other environment variables when defining an environment variable here.
ReFrame will expand its value.
Variables are set after the environment modules are loaded.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.max_jobs
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB8\fP
.UNINDENT
.sp
The maximum number of concurrent regression tests that may be active (i.e., not completed) on this partition.
This option is relevant only when ReFrame executes with the \fI\%asynchronous execution policy\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.prepare_cmds
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
List of shell commands to be emitted before any environment loading commands are emitted.
.sp
New in version 3.5.0.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.resources
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list of job scheduler \fI\%resource specification\fP objects.
.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.processor
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB{}\fP
.UNINDENT
.sp
Processor information for this partition stored in a \fI\%processor info object\fP\&.
If not set, ReFrame will try to auto\-detect this information (see \fI\%Auto\-detecting processor information\fP for more information).
.sp
New in version 3.5.0.

.sp
Changed in version 3.7.0: ReFrame is now able to detect the processor information automatically.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.devices
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
A list with \fI\%device info objects\fP for this partition.
.sp
New in version 3.5.0.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.features
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB[]\fP
.UNINDENT
.sp
User defined features of the partition.
These are accessible through the \fI\%features\fP attribute of the \fI\%current_partition\fP and can also be selected through the extended syntax of \fI\%valid_systems\fP\&.
The values of this list must be alphanumeric strings starting with a non\-digit character and may also contain a \fB\-\fP\&.
.sp
New in version 3.11.0.

.UNINDENT
.INDENT 0.0
.TP
.B \&.systems[]\&.partitions[]\&.extras
.INDENT 7.0
.TP
.B Required
No
.TP
.B Default
\fB{}\fP
.UNINDENT
.sp
User defined attributes of the partition.
These are accessible through the \fI\%extras\fP attribute of the \fI\%current_partition\fP and can also be selected through the extended syntax of \fI\%valid_systems\fP\&.

_old/js/reframe.js
/*
 * @author Victor Holanda Rusu (CSCS)
 *
 */
var reframe =  {};
(function(namespace) {

var __cscsMarkDown = "";

/**
 *  @description reads file contents based on http request. it should be used by all markdown based pages
 *  @param {string} filename
 *  @param {function} callback
 *  @returns {void}
 */
namespace.read_file_contents = function (filename, callback)
{
  var rawFile = new XMLHttpRequest();
  rawFile.open("GET", filename, false);

  var __fileContent = "";
  rawFile.onreadystatechange = function ()
  {
    if(rawFile.readyState === 4) {
      if(rawFile.status === 200 || rawFile.status == 0) {
        __fileContent = rawFile.responseText;
        if(typeof callback == 'function'){
          callback(__fileContent);
        }
      }
    }
  }
  rawFile.send(null);
}

/**
 *  @description gets the base html and parses the sidebar using the same markdown used by remark.js
 *  before changing the markdown one needs to verify if it is compatible with the presentation mode
 *  @param {string} navbarfile
 *  @param {string} sidebarfile
 *  @returns {void}
 */
namespace.setup_site_content = function(navbarfile, sidebarfile, jumbotron) {

  namespace.read_file_contents(navbarfile, function __populate_site_content(argument) {
    document.getElementById("cscs-site-content").innerHTML = argument;
  });

  namespace.read_file_contents(sidebarfile, function cscs_populate_site_content(argument) {
    marked.setOptions({
      gfm: true,
      tables: true,
      breaks: false,
      pedantic: true,
      sanitize: false,
      smartLists: true,
      langPrefix: '',
    });
    marked(argument, function (err, content) {
      if (err) throw err;
      document.getElementById("cscs-leftbar-markdown").innerHTML = content;
    });
    namespace.__correct_navbar();
  });

  if(jumbotron != null) {
    namespace.read_file_contents(jumbotron, function cscs_populate_site_content(argument) {
        marked.setOptions({
          gfm: true,
          tables: true,
          breaks: false,
          pedantic: true,
          sanitize: false,
          smartLists: true,
          langPrefix: '',
        });
        marked(argument, function (err, content) {
          if (err) throw err;
          document.getElementById("cscs-jumbotron-content").innerHTML = content;
        });
      });
  } else {
    $(".jumbotron").hide();
  }

  // presentation mode is hidden if remark.js is not included
  try {
    if(remark != null) {
      $('#start-cscs-presenter-mode').show();
    } else {
      $('#start-cscs-presenter-mode').hide();
    }
  } catch(msg) {
    $('#start-cscs-presenter-mode').hide();
  }

  var presenterMode = document.getElementById('start-cscs-presenter-mode');
  if (presenterMode != null) {
    presenterMode.onclick = namespace.__show_in_presenter_mode;
  }
  namespace.__email_protector();
}

/**
 *  @description this is a wrapper to a selection of functions that need to be called after
 *  the main markdown is rendered
 */
namespace.__correct_navbar = function() {
    $('#cscs-leftbar-markdown').find("a").each(function(index, element) {
        $(element).addClass('reframe-prepend-domain');
    });
    $('#cscs-leftbar-markdown').children("ul").each(function(index, element) {
      if ($(element).hasClass('nav navbar-nav') == false) {
        $(element).addClass('nav navbar-nav');
      }
      $(this).unwrap();
    });

}


/**
 *  @description this is a wrapper to a selection of functions that need to be called after
 *  the main markdown is rendered
 */
namespace.__markdown_post_features = function() {
  namespace.__mouseover_link();
  namespace.__create_toc();
  namespace.__change_table_layout();
  namespace.__highlight_code();
  namespace.__prepend_domain_to_links();
}

/**
 *  @description parses the main website markdown using the same markdown used by remark.js
 *  before changing the markdown one needs to verify if it is compatible with the presentation mode
  * @param {string} navbarfile
  * @param {string} sidebarfile
  * @returns {void}
  */
namespace.setup_markdown_page_content = function(markdownFile) {
  __cscsMarkDown = markdownFile;

  namespace.read_file_contents(markdownFile, function __populate_site_content(argument) {

    marked.setOptions({
      gfm: true,
      tables: true,
      breaks: false,
      pedantic: true,
      sanitize: false,
      smartLists: true,
      langPrefix: '',
    });
    marked(argument, function (err, content) {
      if (err) throw err;
      document.getElementById("cscs-markdown-content").innerHTML = content;
    });
  });
  namespace.__markdown_post_features();
}

/**
 *  @description reads the news markdown and appends it to the cscs-markdown-content.
 *  number_of_news controls the number of news items to print. If negative, all news are printed.
 *  @param {string} news_markdown_file
 *  @param {number} number_of_news
 *  @returns {void}
 */
namespace.read_news = function(news_markdown_file, number_of_news)
{
  namespace.read_file_contents(news_markdown_file, function __populate_site_content(argument) {
    marked(argument, function (err, content) {
      if (err) throw err;

      $('#toc').children().remove();

      var theStart = namespace.__getCommentsObject('#cscs-markdown-content', ' start-news ');
      $(content).insertAfter(theStart);
      var theEnd = namespace.__getCommentsObject('#cscs-markdown-content', ' end-news ');

      if(Number(number_of_news) > 0) {
        var lastOne = $(theEnd).prev();
        if($(theStart).nextUntil(lastOne).filter("h2").length > number_of_news) {
          toRemove = $(theStart).nextUntil(lastOne).filter("h2").slice(0, number_of_news + 1).last();
          toRemove.nextUntil(lastOne).remove();
          toRemove.remove();
          lastOne.remove();
        }
      }
      namespace.__markdown_post_features();
    });
  });
}

/**
 *  @description hides the right toc from the page and makes it a two column page.
 *  @param {string} markdownFile
 *  @returns {void}
 */
namespace.two_column_mode = function(markdownFile) {

  var markdown_div = $('#cscs-markdown-content');
  if (markdown_div.hasClass('col-md-7') == true) {
    markdown_div.removeClass('col-md-7');
  }
  markdown_div.addClass('col-md-9');

  $('#cscs-rightbar').hide();

}

// function cscs_get_modulelist(link, regex, shash_at = 0, elementid = "cscs-markdown-content")
// {
//   read_file_contents(link, function __populate_site_content(argument) {
//     var pattern = regex;
//     var parsed_module = "";

//     var result = pattern.exec(argument);
//     while (result) {
//       var holder = result + '';
//       holder = holder.replace('.eb', '');
//       var splitter = holder.split('-');
//       holder = "";
//       cat = '-';
//       for (var i = 0; i < splitter.length; i++) {
//         if(i == shash_at) {
//           cat = '/';
//         } else if(i == splitter.length -1){
//           cat = '';
//         } else {
//           cat = '-';
//         }
//         holder += splitter[i] + cat;
//       }
//       parsed_module += holder + '\n';
//       result = pattern.exec(argument);
//     }
//     document.getElementById(elementid).innerHTML += parsed_module;
//   });
// }

/**
 *  @description starts the presentation mode. this mode destroys completely the page.
 *  So we refresh the page if the exit button is clicked.
 *  @returns {void}
 */
namespace.__show_in_presenter_mode= function() {
  document.getElementById("cscs-body-container").innerHTML = null;

  try {
    var slideshow = remark.create({sourceUrl: __cscsMarkDown});

    // the click blocks, so forcing full page reload for the click
    var presenterMode = document.getElementById('start-cscs-presenter-mode');
    presenterMode.style.display = 'none';
    var exitMode = document.getElementById('exit-cscs-presenter-mode');
    exitMode.style.display = 'block';

    // workaround to work for local and non local servers
    if(document.location.domain == null)
      exitMode.href = document.location.pathname;
    else
      exitMode.href = document.location.domain+document.location.pathname;

    slideshow.on();
  } catch(msg) {
    $('#start-cscs-presenter-mode').hide();
  }
}

/**
 *  @description protects cscs' e-mail from robots
 *  @returns {void}
 */
namespace.__email_protector = function() {
  // $("#cscs-email-protector").prepend('<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;%69%6E%66%6F%40%63%73%63%73%2E%63%68">Contact CSCS</a>');
  $("#cscs-email-protector").prepend('<a href="&#109;&#097;&#105;&#108;&#116;&#111;:&#104;&#101;&#108;&#112;&#064;&#099;&#115;&#099;&#115;&#046;&#099;&#104;">Contact us</a>');
}

/**
 *  @description wraps the markdown headers with a link it does it on mouse hover.
 *  But it is not really needed. We could do it immediately without the need of hovering
 *  @returns {void}
 */
namespace.__mouseover_link = function() {

  $('#cscs-markdown-content').children("h1, h2, h3").each(function(index, element) {
    $(element).hover(
      function() {
        $(this).wrap(function() {
          return "<a id='" + $(this).attr('id') + "' href='#" + $( this ).attr('id') + "'></a>";
        });
      },
      function() {
        $(this).unwrap();
      }
    );
  });
}

/**
 *  @description creates the toc based on the main markdown content.
 *  @returns {void}
 */
namespace.__create_toc = function() {
  // TOC creation
  // $('#toc').TOC();
  $('#toc').TOC("#cscs-markdown-content", {
    headings: ["h1","h2"],
  });

  $('#toc').append('<a class="back-to-top" href="#">Back to top</a>');
  // Sidenav affixing
  setTimeout(function () {
    $('#toc').affix({
      offset: {
        // top: $('.cscs-global-nav').outerHeight(),
        // top: function () {
        //   return $('.cscs-global-nav').outerHeight()
        // },
        top: function () {
          return $('.jumbotron').outerHeight()
        },
        bottom: function () {
          return (this.bottom = $('.footer').outerHeight(true));
        }

      }
    })}, 100);
  var $window = $(window);
  var $body   = $(document.body);

  $body.scrollspy({
    target: '#toc'
  });

  $window.on('load', function () {
    $body.scrollspy('refresh');
  });

  // counting the number of h1 levels in the toc
  var tocsize = 0;
  $('#toc').children("ul").each(function(index, element) {
    var size = $(this).children("li").length;
    if(size > tocsize){
      tocsize = size;
    }
  });

  // expanding toc when there is only one h1
  if(tocsize == 1) {
    $("head").append("<style> nav[data-toggle='toc'] .nav .nav { display: block; } </style>");
  }
}

/**
 *  @description destroys the remark presentation and restores the CSCS website. This should be deprecated
 *  @returns {void}
 */
namespace.__exit_presentation_mode = function() {
  if(document.location.domain == null)
    window.location.assign(document.location.pathname);
  else
    window.location.assign(document.location = document.location.domain+document.location.pathname);
}

/**
 *  @description changes the layout of table inside the main side markdown area.
 *  @returns {void}
 */
namespace.__change_table_layout = function() {
  $('#cscs-markdown-content').children("table").each(function(index, element) {
    $(element).addClass('table table-striped table-bordered' );
    $(this).wrap(function() {
      return "<div class='table-responsive' ></a>";
    });
  });
}

/**
 *  @description highlights `pre code` blocks.
 *  @returns {void}
 */
namespace.__highlight_code = function() {
  try {
      $('pre').addClass('line-numbers');

      $('pre code').addClass(function( index, currentClass ) {
        var addedClass = "language-" + currentClass;
        return addedClass;
      });
  } catch(msg) {
      console.log('error');
      console.log(msg);
    // no need to catch
  }
}

/**
 *  @description get the current domain.
 *  @returns {string}
 */
namespace.__correct_domain = function()
{
  // var domain = window.location.origin;
  var domain = window.location.host + "/";
  // var domain = "";

  if(document.location.domain != null) {
    domain = document.location.domain;
  }
  domain += "/";

  // console.log('window.location.origin: ' + window.location.origin);
  // console.log('window.location.host: ' + window.location.host);
  // console.log('window.location.hostname: ' + window.location.hostname);
  // console.log('window.location.pathname: ' + window.location.pathname);
  // console.log('window.location.domain: ' + window.location.domain);

  var folders = [ 'about', 'pipeline', 'running', 'started', 'structure', 'usecases', 'writing_checks', 'configure', 'examples', 'reference', '' ]
  var paths = window.location.pathname.split('/');
  paths.forEach(function(element) {
    // console.log(element);
    if ((folders.indexOf(element) == -1) && (element.indexOf('.') == -1)) {
      domain += element + "/";
    }
  }, this);

  domain = domain.replace('//', '/');
  domain = domain.replace('//', '/');
  domain = domain.replace('//', '/');

  return domain
}

/**
 *  @description prepends the domain to the main content.
 *  @returns {string}
 */
namespace.__prepend_domain_to_links = function()
{
  // var domain = window.location.origin;
  var domain = namespace.__correct_domain();

  // prepending cscs domain
  $('.reframe-prepend-domain').each(function(index, element) {
    $(element).attr('href',function(i,v) {
      return window.location.protocol + '//' + (domain + v).replace('//', '/');
    });
  });

  $('#cscs-leftbar-markdown').find('a').each(function(index, element) {
    $(element).attr('href',function(i, v) {
      return window.location.protocol + '//' + (domain + v).replace('//', '/');
    });
  });

  var ele = $('#cscs-markdown-content');
  ele.find('a').each(function(index, element) {
    $(element).attr('href',function(i, v) {
      if (v.startsWith("/", 0)) {
        return window.location.protocol + '//' + (domain + v).replace('//', '/');
      } else {
        return v;
      }
    });
  });
}

/**
 *  @description helper function to get the position of the comments in the page.
 *  This is important for the news pages.
 *  @returns {void}
 */
namespace.__getCommentsObject = function(element, comment) {
  var returnValue = null;
  $(element).contents().filter(function(){
    return this.nodeType == 8;
  }).each(function(i, e){
    if(e.nodeValue === comment) {
      returnValue = e;
      return false;
    }
  });
  return returnValue;
}
})(reframe);
